### ISSUE TITLE
Route dependency is running not in the same thread 

### ISSUE THREAD

**Issue by vdytyniak-exos** [tiangolo/fastapi#5891]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def configure_database(user_arn: str | None = Header(default=None)):
    database = get_database(user_arn)

    with database.bind_ctx(_BaseModel.subclasses()):
        try:
            database.connect()
            print('deps', user_arn, threading.get_ident())
            yield database
        finally:
            if not database.is_closed():
                database.close()

@router.get('/db_username')
def get_db_username(user_arn: str | None = Header(default=None), db=Depends(configure_database)):
    print('api', user_arn, threading.get_ident())
    current_db_user = db.execute_sql('SELECT current_user;').fetchone()[0]
    return current_db_user
```

### Description

Example is partially inspired by [FastAPI - Peewee integration](https://fastapi.tiangolo.com/advanced/sql-databases-peewee/) doc. In most cases everything works fine, but sometimes  `configure_database` and `get_db_username` route are running in different threads and it leads to issues with database connections. Is it expected that Depends and route can run in different threads?

### Operating System

macOS

### FastAPI Version

0.87.0

### Python Version

3.10.5

**Comment by Kludex** [comment]
As of now, yes. It's expected.

**Comment by vdytyniak-exos** [comment]
Does it mean that this [example](https://fastapi.tiangolo.com/advanced/sql-databases-peewee/#create-a-dependency) can have issues with connection created in one thread and route function logic running in another one?

**Comment by caeser1996** [comment]
Yes, it's possible for the Depends and the route to run in different threads, especially in a multi-threaded environment. It's important to handle thread-safety when using databases in FastAPI, as each thread will have its own database connection.

In your example, the configure_database function is using the yield statement which allows it to act as a generator function, returning an instance of the database for each iteration. Each iteration of the generator function creates a new database connection and is executed in a separate thread.

To resolve the issue, you can use thread-local storage to store the database connection and share it between the Depends and the route. The threading.local class provides thread-local storage that allows you to store thread-local data. Here's an example of how you can use it in your code:

```
import threading

_local = threading.local()

def get_database():
    if not hasattr(_local, 'database'):
        _local.database = get_database(user_arn)
    return _local.database

def configure_database(user_arn: str | None = Header(default=None)):
    database = get_database()

    with database.bind_ctx(_BaseModel.subclasses()):
        try:
            database.connect()
            print('deps', user_arn, threading.get_ident())
            return database
        finally:
            if not database.is_closed():
                database.close()

@router.get('/db_username')
def get_db_username(user_arn: str | None = Header(default=None), db=Depends(configure_database)):
    print('api', user_arn, threading.get_ident())
    current_db_user = db.execute_sql('SELECT current_user;').fetchone()[0]
    return current_db_user

```
This way, each thread will have its own instance of _local and its own database connection, and the connection will be shared between the Depends and the route in the same thread.
